### DISCUSSION TITLE
Best way to detect when when automatic reconnect will be attempted and when not

### DISCUSSION THREAD

**Discussion by Etuosaa** [doghappy/socket.io-client-csharp#271] [Q&A]
Hi,

I need to keep attempting reconnection in both cases, when connection was lost and after server closes the connection. What would be the best way to know when automatic reconnection won't be attempted (so that I can manually connect instead)? It seems that calling ConnectAsync again when automatic reconnection is happening might lead to multiple sockets being opened.

Thank you!

**Comment by doghappy** [comment]
there is a `ReconnectionAttempts` option, default value is `int.MaxValue`. so the maximum number of times the library will try to connect is `2147483647`

there is an event: OnReconnectFailed. it will fired when couldn’t reconnect within reconnectionAttempts

for these 2 reasons, it will not automatically reconnect
- io server disconnect: call `disconnect` from server
- io client disconnect: call `disconnect` from client

```cs
private static void Socket_OnDisconnected(object sender, string e)
{
    if (e == DisconnectReason.IOServerDisconnect)
    {
        // the disconnection was initiated by the server, you need to reconnect manually
        var io = sender as SocketIO;
        io.connect();
    }
}
```

**Reply by Etuosaa** [reply]
Thanks,

Partly related, I noticed that OnDisconnected can get called before the transport is disconnected. I was reconnecting in OnDisconnected, but that seemed to occasionally lead to a couple disconnect/reconnects. Delaying reconnect a bit helped at least in most cases. Automatic reconnect seems to be called after transport has disconnected.

Is this by design, and is there a better option to reconnect on any disconnect? I'm delaying the reconnect by a lengthy amount now, which seems to work but is a bit hacky.
